README.markdown
# FileTransferSenderSubdirectoriesPlugin

An Adium plugin that puts incoming files in per-sender subdirectories of the default directory.

A compiled version can be installed [from Adium Xtras](http://www.adiumxtras.com/index.php?a=xtras&xtra_id=4282).

## Functionality

If your Adium is configured to automatically accept files, things should Just Work. If you're prompted for the download directory, going with the default directory (the "Save files to" directory in the File Transfer settings) will cause the file(s) to go into a sender subdirectory. If any directory other than the default directory is selected, the file(s) will go where specified and not into a subdirectory.

The subdirectory will be named "Display Name (Unique Name)", where the "Unique Name" is unchanging: the ICQ UIN, MSN username, Jabber JID etc. The same directory will be used for later file transfers as long as the unique name is unchanged.

## Building

In addition to the code in this repository, you need to add the linked frameworks

  * Adium.framework
  * AIUtilities.framework
  * FriBidi.framework

to the project directory. Compile Adium to get compiled versions of these frameworks in build/{Deployment|Development}.

You can read more about [checking out the Adium source](http://trac.adiumx.com/wiki/GettingReleaseAdiumSource) and [writing Adium plugins](http://fadeover.org/blog/archives/25).

## Credits

By [Henrik Nyh](http://henrik.nyh.se/).


CHANGELOG.txt
1.1 (2008-08-28): Adium 1.3 compatibility.
1.0 (2007-05-24): First version.


FileTransferFoldersPerSenderPlugin.m
//
//  FileTransferFoldersPerSenderPlugin
//
//  By Henrik Nyh, 2007-05-24.
//  Free to modify and redistribute with due credit.
//

#import "FileTransferFoldersPerSenderPlugin.h"
#import <Adium/AIAdiumProtocol.h>
#import <Adium/AIPreferenceControllerProtocol.h>
#import <Adium/ESFileTransfer.h>
#import <AIUtilities/AIFileManagerAdditions.h>
#import <AIUtilities/AIStringAdditions.h>

@implementation FileTransferFoldersPerSenderPlugin

- (NSString *)pluginAuthor {
	return @"Henrik Nyh";
}
- (NSString *)pluginURL {
	return @"http://henrik.nyh.se/";
}
- (NSString *)pluginVersion {
	return @"1.0";
}
- (NSString *)pluginDescription {
	return @"Puts incoming files in per-sender subdirectories of the download folder.";
}


- (void)installPlugin {

	NSLog(@"FileTransferFoldersPerSenderPlugin loaded!");
	[[adium notificationCenter] addObserver:self
	                               selector:@selector(transferStarted:)
	                                   name:FILE_TRANSFER_BEGAN
	                                 object:nil];
}


- (void)uninstallPlugin {

	[[adium notificationCenter] removeObserver:self];
	NSLog(@"FileTransferFoldersPerSenderPlugin unloaded!");
}


- (void)transferStarted:(NSNotification *)notification {

	ESFileTransfer *transfer = (ESFileTransfer *)[notification userInfo];
	
	// Determine the name for the user folder: "Display Name (User ID)" or "User ID"
	NSString *displayName = [[transfer contact] displayName];
	NSString *fUID = [[transfer contact] formattedUID];
	NSString *userFolderName = [NSString stringWithFormat: (displayName ? @"%@ (%@)" : @"%@%@"), displayName, fUID];

	// Replace "/" with "-" and abbreviate long names
	userFolderName = [userFolderName safeFilenameString];

	// Figure out where this file is destined
	NSString *destinationPath = [transfer localFilename];
	NSString *destinationFolder = [destinationPath stringByDeletingLastPathComponent];

	// Use the remote filename, since the local filename may have been uniqued ("foo.jpg" becomes "foo-1.jpg") based on the contents of the default download folder rather than the user folder
	NSString *destinationFile = [transfer remoteFilename];
	
	// Bail if the file is not destined for the default download folder (i.e. the user was prompted and chose someplace else)
	NSString *defaultFolder = [[adium preferenceController] userPreferredDownloadFolder];
	if (![destinationFolder isEqualToString:defaultFolder]) return;
	
	NSFileManager *fileManager = [NSFileManager defaultManager];

	// Find existing user folder, if any: the UID should be identical, though the display name may vary
	NSString *fileOrDir;
	NSDirectoryEnumerator *dirEnumerator = [fileManager enumeratorAtPath:defaultFolder];
	while (fileOrDir = [dirEnumerator nextObject]) {
		BOOL sameUID = [fileOrDir isEqualToString:fUID] || [fileOrDir hasSuffix:[NSString stringWithFormat:@" (%@)", fUID]];
		BOOL isDirectory; [fileManager fileExistsAtPath:[defaultFolder stringByAppendingPathComponent:fileOrDir] isDirectory:&isDirectory];
		if (sameUID && isDirectory) {
			userFolderName = fileOrDir;
			break;
		}
	}
	
	// Create user folder if necessary
	NSString *userFolder = [defaultFolder stringByAppendingPathComponent:userFolderName];
	if (![fileManager fileExistsAtPath:userFolder]) {
		[fileManager createDirectoryAtPath:userFolder attributes:nil];
	}

	// Change destination filename
	NSString *userPath = [fileManager uniquePathForPath:[userFolder stringByAppendingPathComponent:destinationFile]];
	[transfer setLocalFilename:userPath];
}

@end


FileTransferSenderSubdirectoriesPlugin.m
//
//  FileTransferSenderSubdirectoriesPlugin
//
//  By Henrik Nyh, 2007-05-24.
//  Free to modify and redistribute with due credit.
//

#import "FileTransferSenderSubdirectoriesPlugin.h"
#import <Adium/AIAdiumProtocol.h>
#import <Adium/AIPreferenceControllerProtocol.h>
#import <Adium/ESFileTransfer.h>
#import <AIUtilities/AIFileManagerAdditions.h>
#import <AIUtilities/AIStringAdditions.h>

@implementation FileTransferSenderSubdirectoriesPlugin

- (NSString *)pluginAuthor {
	return @"Henrik Nyh";
}
- (NSString *)pluginURL {
	return @"http://henrik.nyh.se/";
}
- (NSString *)pluginVersion {
	return @"1.1";
}
- (NSString *)pluginDescription {
	return @"Puts incoming files in per-sender subdirectories of the default directory.";
}


- (void)installPlugin {

	NSLog(@"FileTransferSenderSubdirectoriesPlugin loaded!");
	[[adium notificationCenter] addObserver:self
	                               selector:@selector(transferStarted:)
	                                   name:FILE_TRANSFER_BEGAN
	                                 object:nil];
}


- (void)uninstallPlugin {

	[[adium notificationCenter] removeObserver:self];
	NSLog(@"FileTransferSenderSubdirectoriesPlugin unloaded!");
}


- (void)transferStarted:(NSNotification *)notification {

	ESFileTransfer *transfer = (ESFileTransfer *)[notification userInfo];
	
	// Determine the name for the user directory: "Display Name (User ID)" or "User ID"
	NSString *displayName = [[transfer contact] displayName];
	NSString *fUID = [[transfer contact] formattedUID];
	NSString *userDirectoryName = [NSString stringWithFormat: (displayName ? @"%@ (%@)" : @"%@%@"), displayName, fUID];

	// Replace "/" with "-" and abbreviate long names
	userDirectoryName = [userDirectoryName safeFilenameString];

	// Figure out where this file is destined
	NSString *destinationPath = [transfer localFilename];
	NSString *destinationDirectory = [destinationPath stringByDeletingLastPathComponent];

	// Use the remote filename, since the local filename may have been uniqued ("foo.jpg" becomes "foo-1.jpg") based on the contents of the default download directory rather than the user directory
	NSString *destinationFile = [transfer remoteFilename];
	
	// Bail if the file is not destined for the default download directory (i.e. the user was prompted and chose someplace else)
	NSString *defaultDirectory = [[adium preferenceController] userPreferredDownloadFolder];
	if (![destinationDirectory isEqualToString:defaultDirectory]) return;
	
	NSFileManager *fileManager = [NSFileManager defaultManager];

	// Find existing user directory, if any: the UID should be identical, though the display name may vary
	NSString *fileOrDir;
	NSDirectoryEnumerator *dirEnumerator = [fileManager enumeratorAtPath:defaultDirectory];
	while (fileOrDir = [dirEnumerator nextObject]) {
		BOOL sameUID = [fileOrDir isEqualToString:fUID] || [fileOrDir hasSuffix:[NSString stringWithFormat:@" (%@)", fUID]];
		BOOL isDirectory; [fileManager fileExistsAtPath:[defaultDirectory stringByAppendingPathComponent:fileOrDir] isDirectory:&isDirectory];
		if (sameUID && isDirectory) {
			userDirectoryName = fileOrDir;
			break;
		}
	}
	
	// Create user directory if necessary
	NSString *userDirectory = [defaultDirectory stringByAppendingPathComponent:userDirectoryName];
	if (![fileManager fileExistsAtPath:userDirectory]) {
		[fileManager createDirectoryAtPath:userDirectory attributes:nil];
	}

	// Change destination filename
	NSString *userPath = [fileManager uniquePathForPath:[userDirectory stringByAppendingPathComponent:destinationFile]];
	[transfer setLocalFilename:userPath];
}

@end
